src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs

```diff
@@ -58,6 +58,13 @@ public CommonCompiler(CommandLineParser parser, string responseFile, string[] ar
             this.Arguments = parser.Parse(allArgs, baseDirectory, sdkDirectoryOpt, additionalReferenceDirectories);
             this.MessageProvider = parser.MessageProvider;
             this.AnalyzerLoader = analyzerLoader;
+
+#if DEBUG
+            if (Arguments.ParseOptions.Features.ContainsKey("debug-determinism"))
+            {
+                EmitDeterminismKey(Arguments, args, baseDirectory, parser);
+            }
+#endif
```

Comment:
This is a block that I'm interested in getting feedback on.  

From personal experience I can attest that it's nearly impossible to diagnose deterministic build issues without a diagnostic file of this nature.  There are just too many subtle factors that can affect determinism and it's impossible to verify them all in the debugger.  Having a file get emitted to disk which captures a summary of the build makes it super easy to compare builds: just diff the diagnostic file.  

Eventually I'd like to have this be a feature for customers to consume as well.  I'm not sure this is the best approach though.